### ISSUE TITLE
H3: still previews OOM during VAE decode on a 16 GB card, then previews are disabled for the rest of the run

### ISSUE THREAD

**Issue by Hell-Bent-Fox** [shootthesound/Fizgig#134]
On a 16 GB card, MiniMax H3 LoRA training with in-training previews (stills) fails every preview with OOM when the run moves from the sampling phase to the VAE decode phase. After the second failure the trainer latches previews off for the whole run:

    WARNING:fizgig.minimax.trainer:[preview] Sample at Start failed (OutOfMemoryError) — training continues; per-epoch previews will still be attempted.
    ...
    WARNING:fizgig.minimax.trainer:[preview] epoch 1 preview failed (CUDA OOM); disabling previews for the rest of the run. Training continues and LoRAs still save normally.

## Environment

- Fizgig v5.7.0 (same behaviour on v5.3.3)
- Linux, 16 GB CUDA card (driver reports 15.77 GB)
- MiniMax H3 base, int8
- Dataset: 32 photos, 50 epochs, batch 1, buckets up to 672×352
- Previews: still (1 frame), 640×512, 20 steps, seed 1234, every epoch, full VAE decode
- Optimised Likeness on, EMA on (0.98), caption dropout 0.05

## VRAM plan

    INFO:fizgig.minimax.trainer:[vram] auto plan: free 16.6 GB, largest bucket 0.25 MP, base ~21 GB (int8, pruned), adapter ~0.7 GB (37 M params, adamw) +0.16 GB frozen LoRAs +0.15 GB EMA shadow -> blocks_to_swap=26, checkpointing=on
    INFO:fizgig.minimax.trainer:[vram] base precision: int8 — int8 with 26 blocks streamed H2D-only — the accurate base, and streaming (not parking) keeps the swap cheap
    INFO:fizgig.minimax.trainer:[vram] block swap active: last 26 blocks streamed H2D-only (int8, ring 2, ~10.0 GB pinned in RAM) — no writeback, prefetch overlaps compute

## Epoch 0 preview — sampling completes, decode OOMs

All 20 sampling steps run fine (4.1 s each). The fp16 video VAE decoder is ~4.8 GB, and at the pre-decode point the card has only 3.96 GB free:

    INFO:fizgig.minimax.trainer:[preview] training adapter off for the render (deployment view) — parked on the CPU (0.14 GB back to the card, 4.4 GB free now)
    [preview] epoch 0: prompt 1/1 (640x512, 1 frame(s), seed 1234)
    [preview] step 1/20  sigma 1.0000 -> 0.9956
    [preview] step 2/20  sigma 0.9956 -> 0.9908  (4.1s)
    ...
    [preview] step 20/20  sigma 0.3871 -> 0.0000  (4.1s)
    INFO:fizgig.minimax.trainer:[vram:pre-decode] allocated 11.35 / reserved 11.39 GB (inactive-split 0.00), driver free 3.96 of 15.77 GB
    INFO:fizgig.minimax.trainer:[vram:finally-done] allocated 11.35 / reserved 11.39 GB (inactive-split 0.00), driver free 3.96 of 15.77 GB
    WARNING:fizgig.minimax.trainer:[preview] Sample at Start failed (OutOfMemoryError) — training continues; per-epoch previews will still be attempted.

## Epoch 1 preview — only 0.53 GB left, OOM again, previews disabled for the run

    INFO:fizgig.minimax.trainer:[preview] training adapter off for the render (deployment view) — parked on the CPU (0.14 GB back to the card, 0.7 GB free now)
    [preview] epoch 1: prompt 1/1 (640x512, 1 frame(s), seed 1234)
    [preview] step 1/20  sigma 1.0000 -> 0.9956
    [preview] step 2/20  sigma 0.9956 -> 0.9908  (4.0s)
    ...
    [preview] step 20/20  sigma 0.3871 -> 0.0000  (4.1s)
    INFO:fizgig.minimax.trainer:[vram:pre-decode] allocated 11.45 / reserved 11.49 GB (inactive-split 0.00), driver free 0.53 of 15.77 GB
    INFO:fizgig.minimax.trainer:[preview] 0.6 GB free after the preview — defragmenting via a full park/restore round-trip so the next training step doesn't fight the fragments.
    INFO:fizgig.minimax.trainer:[vram:post-defrag] allocated 11.45 / reserved 11.47 GB (inactive-split 0.00), driver free 0.53 of 15.77 GB
    INFO:fizgig.minimax.trainer:[vram:finally-done] allocated 11.45 / reserved 11.47 GB (inactive-split 0.00), driver free 0.53 of 15.77 GB
    WARNING:fizgig.minimax.trainer:[preview] epoch 1 preview failed (CUDA OOM); disabling previews for the rest of the run. Training continues and LoRAs still save normally.

## Observations

1. **A still preview has no decode headroom.** The decoder is ~4.8 GB; pre-decode free is 3.96 GB (epoch 0) and 0.53 GB (epoch 1). The sampling phase fits, but the decode phase cannot even load the decoder. Note the pre-decode tail-block parking guard runs only when the preview is a clip (`_frames > 1`); for stills the decoder is loaded straight onto the card.

2. **~3.3 GB of driver-side free VRAM disappears after the first training epoch.** Between the epoch 0 preview and the epoch 1 preview, driver free fell 3.96 GB → 0.53 GB while torch reserved grew only 11.39 → 11.49 GB (+0.10 GB). The missing ~3.3 GB sits on the driver side, outside the PyTorch allocator's accounting. Training steps keep running normally (~9 s/it) — only the preview decode is starved.

3. For context, the first training steps of the run were slow (129 s → 162 s → 92 s → 60 s → 42 s → 30 s → 24 s → settling at ~9 s/it after ~8 steps), consistent with Triton kernel autotuning on first use — the same window in which the driver-side memory above appeared.

## Impact

On a 16 GB int8 plan, in-training still previews fail at epoch 0 and epoch 1, and the run then silently loses its per-epoch previews for the remaining 48 epochs — the only in-run signal of what the LoRA is learning.

## Opinion

Perhaps it’s worth considering the correction option I already proposed earlier. But perhaps the problem is only with me, and then I’ll take the necessary steps myself.

**Comment by shootthesound** [comment]
You were right in August and you are right now. #109 had exactly this fix. The work I said was in hand never materialised; it got delayed behind other things and I completely forgot to come back to it, which is on me. I've reopened #109 and merged it as yours, so it goes out in the next release with your name on it. The 3.3 GB that vanishes outside PyTorch after the first epoch is a separate oddity worth its own look, but with the parking now driven by the live free figure the decode should fit either way. Thank you for coming back with the log.
